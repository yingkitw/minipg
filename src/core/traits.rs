//! Core traits for capability-facing abstractions.

use super::{Diagnostic, Result};
use std::path::Path;

/// Trait for parsing grammar files.
pub trait GrammarParser {
    type Output;

    /// Parse a grammar from a file.
    fn parse_file(&self, path: &Path) -> Result<Self::Output>;

    /// Parse a grammar from a string.
    fn parse_string(&self, source: &str, filename: &str) -> Result<Self::Output>;
}

/// Trait for semantic analysis of grammars.
pub trait SemanticAnalyzer {
    type Input;
    type Output;

    /// Perform semantic analysis on the input.
    fn analyze(&self, input: &Self::Input) -> Result<Self::Output>;

    /// Get diagnostics from the analysis.
    fn diagnostics(&self) -> &[Diagnostic];
}

/// Trait for code generation.
pub trait CodeGenerator {
    type Input;
    type Config;

    /// Generate code for the given input and configuration.
    fn generate(&self, input: &Self::Input, config: &Self::Config) -> Result<String>;

    /// Get the target language name.
    fn target_language(&self) -> &str;
}

/// Trait for grammar transformation.
pub trait GrammarTransformer {
    type Input;
    type Output;

    /// Transform the grammar.
    fn transform(&self, input: Self::Input) -> Result<Self::Output>;
}

/// Trait for error reporting.
pub trait ErrorReporter {
    /// Report a diagnostic.
    fn report(&mut self, diagnostic: Diagnostic);

    /// Get all reported diagnostics.
    fn diagnostics(&self) -> &[Diagnostic];

    /// Check if any errors have been reported.
    fn has_errors(&self) -> bool;

    /// Get the number of errors.
    fn error_count(&self) -> usize;

    /// Get the number of warnings.
    fn warning_count(&self) -> usize;
}

/// Trait for grammar validation.
pub trait GrammarValidator {
    type Input;

    /// Validate the grammar.
    fn validate(&self, input: &Self::Input) -> Result<()>;
}
