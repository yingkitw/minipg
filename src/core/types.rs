//! Common types used across minipg.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Grammar type.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum GrammarType {
    Lexer,
    Parser,
    Combined,
}

/// Token type identifier.
pub type TokenType = usize;

/// Rule identifier.
pub type RuleId = usize;

/// Symbol table for tracking names and their types.
#[derive(Debug, Clone, Default)]
pub struct SymbolTable {
    tokens: HashMap<String, TokenType>,
    rules: HashMap<String, RuleId>,
}

impl SymbolTable {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_token(&mut self, name: String, token_type: TokenType) {
        self.tokens.insert(name, token_type);
    }

    pub fn add_rule(&mut self, name: String, rule_id: RuleId) {
        self.rules.insert(name, rule_id);
    }

    pub fn get_token(&self, name: &str) -> Option<TokenType> {
        self.tokens.get(name).copied()
    }

    pub fn get_rule(&self, name: &str) -> Option<RuleId> {
        self.rules.get(name).copied()
    }

    pub fn tokens(&self) -> &HashMap<String, TokenType> {
        &self.tokens
    }

    pub fn rules(&self) -> &HashMap<String, RuleId> {
        &self.rules
    }
}

/// Configuration options for code generation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeGenConfig {
    pub target_language: String,
    pub output_directory: String,
    pub package_name: Option<String>,
    pub generate_listener: bool,
    pub generate_visitor: bool,
}

impl Default for CodeGenConfig {
    fn default() -> Self {
        Self {
            target_language: "rust".to_string(),
            output_directory: ".".to_string(),
            package_name: None,
            generate_listener: true,
            generate_visitor: false,
        }
    }
}
