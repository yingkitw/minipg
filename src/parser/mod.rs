//! Grammar parser implementation.

pub mod lexer;
pub mod parser;
pub mod token;
pub mod error_recovery;

pub use lexer::Lexer;
pub use parser::Parser;
pub use token::{Token, TokenKind};

use crate::ast::Grammar;
use crate::core::{GrammarParser as GrammarParserTrait, Result};
use std::path::Path;

/// Main grammar parser.
pub struct GrammarParser;

impl GrammarParser {
    pub fn new() -> Self {
        Self
    }
}

impl Default for GrammarParser {
    fn default() -> Self {
        Self::new()
    }
}

impl GrammarParserTrait for GrammarParser {
    type Output = Grammar;

    fn parse_file(&self, path: &Path) -> Result<Self::Output> {
        let source = std::fs::read_to_string(path)?;
        let filename = path.to_string_lossy().to_string();
        self.parse_string(&source, &filename)
    }

    fn parse_string(&self, source: &str, filename: &str) -> Result<Self::Output> {
        let lexer = Lexer::new(source, filename);
        let mut parser = Parser::new(lexer);
        parser.parse()
    }
}
