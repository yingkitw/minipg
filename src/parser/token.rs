//! Token definitions for the grammar lexer.

use std::fmt;

/// Token kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    // Keywords
    Grammar,
    Lexer,
    Parser,
    Fragment,
    Import,
    Options,

    // Identifiers and literals
    Identifier,
    StringLiteral,
    CharLiteral,

    // Operators and punctuation
    Colon,
    Semicolon,
    Pipe,
    Comma,
    Question,
    Star,
    Plus,
    Dot,
    Range,
    Not,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Equals,
    Arrow,

    // Special
    Eof,
    Error,
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TokenKind::Grammar => write!(f, "grammar"),
            TokenKind::Lexer => write!(f, "lexer"),
            TokenKind::Parser => write!(f, "parser"),
            TokenKind::Fragment => write!(f, "fragment"),
            TokenKind::Import => write!(f, "import"),
            TokenKind::Options => write!(f, "options"),
            TokenKind::Identifier => write!(f, "identifier"),
            TokenKind::StringLiteral => write!(f, "string literal"),
            TokenKind::CharLiteral => write!(f, "char literal"),
            TokenKind::Colon => write!(f, ":"),
            TokenKind::Semicolon => write!(f, ";"),
            TokenKind::Pipe => write!(f, "|"),
            TokenKind::Comma => write!(f, ","),
            TokenKind::Question => write!(f, "?"),
            TokenKind::Star => write!(f, "*"),
            TokenKind::Plus => write!(f, "+"),
            TokenKind::Dot => write!(f, "."),
            TokenKind::Range => write!(f, ".."),
            TokenKind::Not => write!(f, "~"),
            TokenKind::LeftParen => write!(f, "("),
            TokenKind::RightParen => write!(f, ")"),
            TokenKind::LeftBrace => write!(f, "{{"),
            TokenKind::RightBrace => write!(f, "}}"),
            TokenKind::LeftBracket => write!(f, "["),
            TokenKind::RightBracket => write!(f, "]"),
            TokenKind::Equals => write!(f, "="),
            TokenKind::Arrow => write!(f, "->"),
            TokenKind::Eof => write!(f, "end of file"),
            TokenKind::Error => write!(f, "error"),
        }
    }
}

/// A token with location information.
#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(kind: TokenKind, text: String, line: usize, column: usize) -> Self {
        Self {
            kind,
            text,
            line,
            column,
        }
    }

    pub fn eof(line: usize, column: usize) -> Self {
        Self::new(TokenKind::Eof, String::new(), line, column)
    }

    pub fn error(text: String, line: usize, column: usize) -> Self {
        Self::new(TokenKind::Error, text, line, column)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token_kind_equality() {
        assert_eq!(TokenKind::Grammar, TokenKind::Grammar);
        assert_ne!(TokenKind::Grammar, TokenKind::Lexer);
    }

    #[test]
    fn test_token_kind_display() {
        assert_eq!(TokenKind::Grammar.to_string(), "grammar");
        assert_eq!(TokenKind::Identifier.to_string(), "identifier");
        assert_eq!(TokenKind::Colon.to_string(), ":");
        assert_eq!(TokenKind::Arrow.to_string(), "->");
        assert_eq!(TokenKind::Eof.to_string(), "end of file");
    }

    #[test]
    fn test_token_new() {
        let token = Token::new(TokenKind::Identifier, "test".to_string(), 1, 5);
        assert_eq!(token.kind, TokenKind::Identifier);
        assert_eq!(token.text, "test");
        assert_eq!(token.line, 1);
        assert_eq!(token.column, 5);
    }

    #[test]
    fn test_token_eof() {
        let token = Token::eof(10, 20);
        assert_eq!(token.kind, TokenKind::Eof);
        assert_eq!(token.text, "");
        assert_eq!(token.line, 10);
        assert_eq!(token.column, 20);
    }

    #[test]
    fn test_token_error() {
        let token = Token::error("unexpected char".to_string(), 5, 10);
        assert_eq!(token.kind, TokenKind::Error);
        assert_eq!(token.text, "unexpected char");
        assert_eq!(token.line, 5);
        assert_eq!(token.column, 10);
    }
}
