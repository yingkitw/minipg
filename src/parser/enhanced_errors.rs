//! Enhanced error messages for the parser.
//!
//! This module provides utilities for generating better error messages
//! with context, expected tokens, and helpful suggestions.

use super::token::{Token, TokenKind};
use crate::core::Error;

/// Context information for error reporting
#[derive(Debug, Clone)]
pub struct ErrorContext {
    pub line: usize,
    pub column: usize,
    pub filename: Option<String>,
    pub surrounding_text: Option<String>,
    pub expected: Vec<TokenKind>,
    pub found: Option<TokenKind>,
    pub found_text: Option<String>,
    pub rule_context: Option<String>,
    pub suggestion: Option<String>,
}

impl ErrorContext {
    pub fn new(line: usize, column: usize) -> Self {
        Self {
            line,
            column,
            filename: None,
            surrounding_text: None,
            expected: Vec::new(),
            found: None,
            found_text: None,
            rule_context: None,
            suggestion: None,
        }
    }

    pub fn with_expected(mut self, expected: Vec<TokenKind>) -> Self {
        self.expected = expected;
        self
    }

    pub fn with_found(mut self, found: TokenKind, text: Option<String>) -> Self {
        self.found = Some(found);
        self.found_text = text;
        self
    }

    pub fn with_context(mut self, rule: String, surrounding: String) -> Self {
        self.rule_context = Some(rule);
        self.surrounding_text = Some(surrounding);
        self
    }

    pub fn with_filename(mut self, filename: String) -> Self {
        self.filename = Some(filename);
        self
    }

    pub fn generate_message(&self) -> String {
        let mut msg = String::new();

        // Add location
        let location = if let Some(ref filename) = self.filename {
            format!("{}:{}:{}", filename, self.line, self.column)
        } else {
            format!("{}:{}", self.line, self.column)
        };

        // Add rule context if available
        if let Some(ref rule) = self.rule_context {
            msg.push_str(&format!("In rule '{}': ", rule));
        }

        // Main error message
        if !self.expected.is_empty() {
            if self.expected.len() == 1 {
                msg.push_str(&format!("Expected {}, ", self.expected[0]));
            } else {
                let expected_str = self.expected.iter()
                    .map(|t| t.to_string())
                    .collect::<Vec<_>>()
                    .join(", ");
                msg.push_str(&format!("Expected one of: {}, ", expected_str));
            }
        }

        if let Some(ref found) = self.found {
            if let Some(ref found_text) = self.found_text {
                msg.push_str(&format!("but found {} ('{}')", found, found_text));
            } else {
                msg.push_str(&format!("but found {}", found));
            }
        } else {
            msg.push_str("but found unexpected token");
        }

        // Add surrounding text if available
        if let Some(ref surrounding) = self.surrounding_text {
            msg.push_str(&format!("\n  Context: {}", surrounding));
        }

        // Add suggestion if available
        if let Some(ref suggestion) = self.suggestion {
            msg.push_str(&format!("\n  Suggestion: {}", suggestion));
        }

        format!("Parse error at {}: {}", location, msg)
    }
}

/// Enhanced error with context
pub fn create_enhanced_error(
    token: &Token,
    expected: Vec<TokenKind>,
    rule_context: Option<&str>,
    source: Option<&str>,
) -> Error {
    let mut context = ErrorContext::new(token.line, token.column)
        .with_expected(expected)
        .with_found(token.kind, Some(token.text.clone()));

    if let Some(rule) = rule_context {
        let surrounding = extract_surrounding_text(token, source);
        context = context.with_context(rule.to_string(), surrounding);
    }

    // Generate suggestion
    let suggestion = generate_suggestion(&context);
    let mut ctx = context;
    if let Some(sugg) = suggestion {
        ctx.suggestion = Some(sugg);
    }

    Error::parse(
        format!("{}:{}", ctx.line, ctx.column),
        ctx.generate_message(),
    )
}

/// Extract surrounding text for error context
fn extract_surrounding_text(token: &Token, source: Option<&str>) -> String {
    if let Some(source_text) = source {
        let lines: Vec<&str> = source_text.lines().collect();
        if token.line > 0 && token.line <= lines.len() {
            let line_text = lines[token.line - 1];
            let start = if token.column > 20 {
                token.column - 20
            } else {
                0
            };
            let end = (token.column + 20).min(line_text.len());
            let snippet = &line_text[start..end];
            
            // Show the line with a caret pointing to the error
            let caret_pos = token.column - start;
            let caret = format!("{}^", " ".repeat(caret_pos));
            format!("{}\n{}", snippet, caret)
        } else {
            format!("line {}", token.line)
        }
    } else {
        format!("line {}, column {}", token.line, token.column)
    }
}

/// Generate helpful suggestions based on error context
fn generate_suggestion(context: &ErrorContext) -> Option<String> {
    // Common typos and suggestions
    if let Some(ref found) = context.found {
        match found {
            TokenKind::Identifier if context.expected.contains(&TokenKind::Semicolon) => {
                Some("Did you forget a semicolon?".to_string())
            }
            TokenKind::Semicolon if context.expected.contains(&TokenKind::RightBrace) => {
                Some("Did you mean '}'?".to_string())
            }
            TokenKind::RightBrace if context.expected.contains(&TokenKind::Semicolon) => {
                Some("Did you mean ';'?".to_string())
            }
            TokenKind::RightParen if context.expected.contains(&TokenKind::RightBrace) => {
                Some("Did you mean '}' instead of ')'?".to_string())
            }
            TokenKind::RightBrace if context.expected.contains(&TokenKind::RightParen) => {
                Some("Did you mean ')' instead of '}'?".to_string())
            }
            _ if context.expected.contains(&TokenKind::Colon) && found == &TokenKind::Equals => {
                Some("Did you mean ':' instead of '='?".to_string())
            }
            _ if context.expected.contains(&TokenKind::Equals) && found == &TokenKind::Colon => {
                Some("Did you mean '=' instead of ':'?".to_string())
            }
            _ => None,
        }
    } else {
        None
    }
}

/// Validate character class ranges
pub fn validate_char_class_range(start: char, end: char) -> Result<(), Error> {
    if start > end {
        return Err(Error::parse(
            format!("character class"),
            format!("Invalid character range: '{}' must be <= '{}'", start, end),
        ));
    }

    // Check for valid Unicode ranges
    if !start.is_ascii() || !end.is_ascii() {
        // For non-ASCII, ensure valid Unicode scalar values
        if !char::from_u32(start as u32).is_some() || !char::from_u32(end as u32).is_some() {
            return Err(Error::parse(
                format!("character class"),
                format!("Invalid Unicode character in range"),
            ));
        }
    }

    Ok(())
}

/// Validate Unicode escape sequence
pub fn validate_unicode_escape(hex: &str) -> Result<char, Error> {
    if hex.len() != 4 {
        return Err(Error::parse(
            format!("unicode escape"),
            format!("Unicode escape must be 4 hex digits, got {} digits", hex.len()),
        ));
    }

    let code = u32::from_str_radix(hex, 16)
        .map_err(|_| Error::parse(
            format!("unicode escape"),
            format!("Invalid hex digits in unicode escape: \\u{}", hex),
        ))?;

    char::from_u32(code)
        .ok_or_else(|| Error::parse(
            format!("unicode escape"),
            format!("Invalid Unicode code point: U+{:04X}", code),
        ))
}

/// Parse Unicode escape sequences (supports both \uXXXX and \u{XXXXXX})
pub fn parse_unicode_escape(text: &str) -> Result<char, Error> {
    if text.starts_with("\\u{") {
        // Extended Unicode escape: \u{XXXXXX}
        let hex = text.strip_prefix("\\u{")
            .and_then(|s| s.strip_suffix('}'))
            .ok_or_else(|| Error::parse(
                format!("unicode escape"),
                "Unclosed Unicode escape sequence".to_string(),
            ))?;

        if hex.len() > 6 {
            return Err(Error::parse(
                format!("unicode escape"),
                format!("Unicode escape too long: max 6 hex digits, got {}", hex.len()),
            ));
        }

        let code = u32::from_str_radix(hex, 16)
            .map_err(|_| Error::parse(
                format!("unicode escape"),
                format!("Invalid hex digits in unicode escape: \\u{{{}}}", hex),
            ))?;

        char::from_u32(code)
            .ok_or_else(|| Error::parse(
                format!("unicode escape"),
                format!("Invalid Unicode code point: U+{:04X}", code),
            ))
    } else if text.starts_with("\\u") {
        // Standard Unicode escape: \uXXXX
        let hex = &text[2..];
        validate_unicode_escape(hex)
    } else {
        Err(Error::parse(
            format!("unicode escape"),
            format!("Invalid unicode escape format: {}", text),
        ))
    }
}

/// Edge case validation for grammar elements
pub fn validate_grammar_edge_cases(
    token: &Token,
    context: &str,
) -> Result<(), Error> {
    // Check for common edge cases
    
    // Empty alternatives
    if context.contains("||") {
        return Err(Error::parse(
            format!("{}:{}", token.line, token.column),
            "Empty alternative detected. Did you mean to add content between '|'?".to_string(),
        ));
    }

    // Unterminated strings
    if context.contains("'") && !context.matches("'").count() % 2 == 0 {
        return Err(Error::parse(
            format!("{}:{}", token.line, token.column),
            "Unterminated string literal".to_string(),
        ));
    }

    // Unmatched brackets
    let open_braces = context.matches('{').count();
    let close_braces = context.matches('}').count();
    if open_braces != close_braces {
        return Err(Error::parse(
            format!("{}:{}", token.line, token.column),
            format!("Unmatched braces: {} opening, {} closing", open_braces, close_braces),
        ));
    }

    let open_parens = context.matches('(').count();
    let close_parens = context.matches(')').count();
    if open_parens != close_parens {
        return Err(Error::parse(
            format!("{}:{}", token.line, token.column),
            format!("Unmatched parentheses: {} opening, {} closing", open_parens, close_parens),
        ));
    }

    let open_brackets = context.matches('[').count();
    let close_brackets = context.matches(']').count();
    if open_brackets != close_brackets {
        return Err(Error::parse(
            format!("{}:{}", token.line, token.column),
            format!("Unmatched brackets: {} opening, {} closing", open_brackets, close_brackets),
        ));
    }

    Ok(())
}

