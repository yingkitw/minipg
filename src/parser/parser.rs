//! Parser for grammar files.

use super::lexer::Lexer;
use super::token::{Token, TokenKind};
use crate::ast::{Alternative, Element, Grammar, Rule};
use crate::core::{types::GrammarType, Error, Result};

/// Parser for grammar files.
pub struct Parser {
    lexer: Lexer,
    current_token: Token,
    peek_token: Token,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        let current_token = lexer.next_token();
        let peek_token = lexer.next_token();
        Self {
            lexer,
            current_token,
            peek_token,
        }
    }

    pub fn parse(&mut self) -> Result<Grammar> {
        self.parse_grammar()
    }

    fn parse_grammar(&mut self) -> Result<Grammar> {
        // grammar <type> <name> ;
        self.expect(TokenKind::Grammar)?;
        
        let grammar_type = if self.current_token.kind == TokenKind::Lexer {
            self.advance();
            GrammarType::Lexer
        } else if self.current_token.kind == TokenKind::Parser {
            self.advance();
            GrammarType::Parser
        } else {
            GrammarType::Combined
        };

        let name = self.expect_identifier()?;
        self.expect(TokenKind::Semicolon)?;

        let mut grammar = Grammar::new(name, grammar_type);

        // Parse options, imports, and rules
        while self.current_token.kind != TokenKind::Eof {
            if self.current_token.kind == TokenKind::Options {
                self.parse_options(&mut grammar)?;
            } else if self.current_token.kind == TokenKind::Import {
                self.parse_import(&mut grammar)?;
            } else if self.current_token.kind == TokenKind::Identifier {
                let rule = self.parse_rule()?;
                grammar.add_rule(rule);
            } else if self.current_token.kind == TokenKind::Fragment {
                let rule = self.parse_fragment_rule()?;
                grammar.add_rule(rule);
            } else {
                return Err(Error::parse(
                    format!("{}:{}", self.current_token.line, self.current_token.column),
                    format!("unexpected token: {}", self.current_token.kind),
                ));
            }
        }

        Ok(grammar)
    }

    fn parse_options(&mut self, grammar: &mut Grammar) -> Result<()> {
        self.expect(TokenKind::Options)?;
        self.expect(TokenKind::LeftBrace)?;

        while self.current_token.kind != TokenKind::RightBrace {
            let key = self.expect_identifier()?;
            self.expect(TokenKind::Equals)?;
            let value = self.expect_identifier()?;
            self.expect(TokenKind::Semicolon)?;
            grammar.add_option(key, value);
        }

        self.expect(TokenKind::RightBrace)?;
        Ok(())
    }

    fn parse_import(&mut self, grammar: &mut Grammar) -> Result<()> {
        self.expect(TokenKind::Import)?;
        let import_name = self.expect_identifier()?;
        self.expect(TokenKind::Semicolon)?;
        grammar.add_import(import_name);
        Ok(())
    }

    fn parse_rule(&mut self) -> Result<Rule> {
        let name = self.expect_identifier()?;
        
        // Determine if it's a lexer or parser rule based on first character
        let rule_type = if name.chars().next().unwrap().is_uppercase() {
            crate::ast::RuleType::Lexer
        } else {
            crate::ast::RuleType::Parser
        };

        let mut rule = Rule::new(name, rule_type);
        
        self.expect(TokenKind::Colon)?;
        
        // Parse alternatives
        let alt = self.parse_alternative()?;
        rule.add_alternative(alt);
        
        while self.current_token.kind == TokenKind::Pipe {
            self.advance();
            let alt = self.parse_alternative()?;
            rule.add_alternative(alt);
        }
        
        self.expect(TokenKind::Semicolon)?;
        Ok(rule)
    }

    fn parse_fragment_rule(&mut self) -> Result<Rule> {
        self.expect(TokenKind::Fragment)?;
        let mut rule = self.parse_rule()?;
        rule.set_fragment(true);
        Ok(rule)
    }

    fn parse_alternative(&mut self) -> Result<Alternative> {
        let mut alt = Alternative::new();
        
        while !self.is_alternative_end() {
            let element = self.parse_element()?;
            alt.add_element(element);
        }
        
        Ok(alt)
    }

    fn parse_element(&mut self) -> Result<Element> {
        let element = match self.current_token.kind {
            TokenKind::Identifier => {
                let name = self.expect_identifier()?;
                Element::rule_ref(name)
            }
            TokenKind::StringLiteral => {
                let value = self.current_token.text.clone();
                self.advance();
                Element::string_literal(value)
            }
            TokenKind::LeftParen => {
                self.advance();
                let mut alternatives = vec![self.parse_alternative()?];
                
                while self.current_token.kind == TokenKind::Pipe {
                    self.advance();
                    alternatives.push(self.parse_alternative()?);
                }
                
                self.expect(TokenKind::RightParen)?;
                Element::Group { alternatives }
            }
            TokenKind::LeftBracket => {
                self.advance();
                // Parse character class
                let element = self.parse_char_class()?;
                self.expect(TokenKind::RightBracket)?;
                element
            }
            TokenKind::Dot => {
                self.advance();
                Element::Wildcard
            }
            TokenKind::Not => {
                self.advance();
                let element = self.parse_element()?;
                Element::Not {
                    element: Box::new(element),
                }
            }
            _ => {
                return Err(Error::parse(
                    format!("{}:{}", self.current_token.line, self.current_token.column),
                    format!("unexpected token in element: {}", self.current_token.kind),
                ));
            }
        };

        // Handle suffixes (?, *, +)
        let element = match self.current_token.kind {
            TokenKind::Question => {
                self.advance();
                Element::optional(element)
            }
            TokenKind::Star => {
                self.advance();
                Element::zero_or_more(element)
            }
            TokenKind::Plus => {
                self.advance();
                Element::one_or_more(element)
            }
            _ => element,
        };

        Ok(element)
    }

    fn parse_char_class(&mut self) -> Result<Element> {
        // Check for negation
        let negated = if self.current_token.kind == TokenKind::Not {
            self.advance();
            true
        } else {
            false
        };
        
        let mut ranges = Vec::new();
        
        // Parse character class contents
        while self.current_token.kind != TokenKind::RightBracket 
            && self.current_token.kind != TokenKind::Eof {
            
            if self.current_token.kind == TokenKind::StringLiteral 
                || self.current_token.kind == TokenKind::CharLiteral {
                let start_char = self.parse_char_literal()?;
                
                // Check for range
                if self.current_token.kind == TokenKind::Range {
                    self.advance();
                    if self.current_token.kind == TokenKind::StringLiteral 
                        || self.current_token.kind == TokenKind::CharLiteral {
                        let end_char = self.parse_char_literal()?;
                        ranges.push((start_char, end_char));
                    } else {
                        return Err(Error::parse(
                            format!("{}:{}", self.current_token.line, self.current_token.column),
                            "expected character after range operator".to_string(),
                        ));
                    }
                } else {
                    // Single character
                    ranges.push((start_char, start_char));
                }
            } else {
                break;
            }
        }
        
        if ranges.is_empty() {
            return Err(Error::parse(
                format!("{}:{}", self.current_token.line, self.current_token.column),
                "empty character class".to_string(),
            ));
        }
        
        Ok(Element::CharClass { negated, ranges })
    }
    
    fn parse_char_literal(&mut self) -> Result<char> {
        let text = &self.current_token.text;
        let ch = if text.starts_with("\\u") {
            // Unicode escape: \uXXXX
            let hex = &text[2..];
            u32::from_str_radix(hex, 16)
                .ok()
                .and_then(|code| char::from_u32(code))
                .ok_or_else(|| Error::parse(
                    format!("{}:{}", self.current_token.line, self.current_token.column),
                    format!("invalid unicode escape: {}", text),
                ))?
        } else if text.starts_with('\\') && text.len() == 2 {
            // Simple escape sequences
            match text.chars().nth(1).unwrap() {
                'n' => '\n',
                'r' => '\r',
                't' => '\t',
                '\\' => '\\',
                '\'' => '\'',
                '"' => '"',
                c => c,
            }
        } else {
            text.chars().next().unwrap_or('\0')
        };
        
        self.advance();
        Ok(ch)
    }

    fn is_alternative_end(&self) -> bool {
        matches!(
            self.current_token.kind,
            TokenKind::Pipe
                | TokenKind::Semicolon
                | TokenKind::RightParen
                | TokenKind::RightBracket
                | TokenKind::Eof
        )
    }

    fn expect(&mut self, kind: TokenKind) -> Result<()> {
        if self.current_token.kind == kind {
            self.advance();
            Ok(())
        } else {
            Err(Error::parse(
                format!("{}:{}", self.current_token.line, self.current_token.column),
                format!("expected {}, found {}", kind, self.current_token.kind),
            ))
        }
    }

    fn expect_identifier(&mut self) -> Result<String> {
        if self.current_token.kind == TokenKind::Identifier {
            let text = self.current_token.text.clone();
            self.advance();
            Ok(text)
        } else {
            Err(Error::parse(
                format!("{}:{}", self.current_token.line, self.current_token.column),
                format!("expected identifier, found {}", self.current_token.kind),
            ))
        }
    }

    fn advance(&mut self) {
        self.current_token = std::mem::replace(&mut self.peek_token, self.lexer.next_token());
    }
}
