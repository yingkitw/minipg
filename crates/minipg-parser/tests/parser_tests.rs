//! Parser tests with snapshot testing.

use minipg::core::GrammarParser;
use minipg::parser::GrammarParser as Parser;

#[test]
fn test_parse_simple_grammar() {
    let source = r#"
grammar parser Calculator;

expr: term;
term: NUMBER;

NUMBER: DIGIT+;
DIGIT: '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';
"#;

    let parser = Parser::new();
    let result = parser.parse_string(source, "test.g4");
    
    if let Err(ref e) = result {
        eprintln!("Parse error: {}", e);
    }
    assert!(result.is_ok());
    let grammar = result.unwrap();
    
    insta::assert_yaml_snapshot!(grammar);
}

#[test]
fn test_parse_lexer_grammar() {
    let source = r#"
grammar lexer SimpleLexer;

WS: SPACE+;
ID: LETTER;
NUMBER: DIGIT+;

SPACE: ' ' | '\t' | '\r' | '\n';
LETTER: 'a' | 'b' | 'c';
DIGIT: '0' | '1' | '2';
"#;

    let parser = Parser::new();
    let result = parser.parse_string(source, "test.g4");
    
    if let Err(ref e) = result {
        eprintln!("Parse error: {}", e);
    }
    assert!(result.is_ok());
    let grammar = result.unwrap();
    assert_eq!(grammar.name, "SimpleLexer");
}

#[test]
fn test_parse_error() {
    let source = r#"
grammar parser Invalid
"#;

    let parser = Parser::new();
    let result = parser.parse_string(source, "test.g4");
    
    assert!(result.is_err());
}
