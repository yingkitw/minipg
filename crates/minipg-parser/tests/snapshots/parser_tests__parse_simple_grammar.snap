---
source: crates/minipg-parser/tests/parser_tests.rs
expression: grammar
---
name: Calculator
grammar_type: Parser
options: {}
rules:
  - name: expr
    rule_type: Parser
    alternatives:
      - elements:
          - RuleRef:
              name: term
              label: ~
        label: ~
    is_fragment: false
  - name: term
    rule_type: Parser
    alternatives:
      - elements:
          - RuleRef:
              name: NUMBER
              label: ~
        label: ~
    is_fragment: false
  - name: NUMBER
    rule_type: Lexer
    alternatives:
      - elements:
          - OneOrMore:
              element:
                RuleRef:
                  name: DIGIT
                  label: ~
        label: ~
    is_fragment: false
  - name: DIGIT
    rule_type: Lexer
    alternatives:
      - elements:
          - StringLiteral:
              value: "0"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "1"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "2"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "3"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "4"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "5"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "6"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "7"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "8"
              label: ~
        label: ~
      - elements:
          - StringLiteral:
              value: "9"
              label: ~
        label: ~
    is_fragment: false
imports: []
